include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+	template <typename ParameterT>
+	Parameter<ParameterT>& declare(Parameter<ParameterT>& parameter);
+
+	/*
+	 * Undeclare a parameter, remove all callbacks that have been registered using the declare() call
+	 */
+	template <typename ParameterT>
+	bool undeclare(const Parameter<ParameterT>& parameter);
+	bool undeclare(const std::string& name);
+
+	/*
+	 * Return the parameter for a given name, if not found parameter is nullptr
+	 */
+	// Not possible when working with references, maybe only for actual node parameters?
+	// std::shared_ptr<Parameter<ParameterT>> get(const std::string& name);
+	// std::shared_ptr<const Parameter<ParameterT>> get(const std::string& name) const noexcept;
+
+	//
+	// TODO: insert utility functions:  validate(), print, ostream<<, errors(), warnings()...
+	//
+
+private:
+	const rclcpp::Node::SharedPtr node_;
+	const std::string base_name_;
+
+	// Class similar to the one in rosparam_shortcuts/node_parameters.h
+	// implements ROS calls/checks/callbacks for node_parameters_
+	NodeParametersInterface node_parameters_interface_;
+
+	// Store parameters, only node parameters (rclcpp::Parameters) are stored inside NodeConfig
+	std::map<std::string, Parameter&> parameters_;
+	std::map<std::string, std::shared_ptr<NodeParameter>> node_parameters_;
+
+	// Needed? keep track of registered callbacks that are required for linking SetParameters requests or
+	// exposing high-level update changes (invalid/valid states)
+	// std::string<std::string, std::string> registered_node_parameter_callbacks_ids_;
+	// std::string<std::string, std::string> registered_parameter_callbacks_ids_;
+}
+
+struct ExampleConfig : public NodeConfig
+{
+	ExampleConfig(const rclcpp::Node::SharedPtr& node) : NodeConfig(node) {
+		// declare simple parameters just like that
+		declare(int1_param);
+		declare(int2_param);
+		declare(int3_param);
+
+		// "double_param", add some description and validity checks
+		declare(double_param)
+		    .describe("This is a very important double parameter", "should be greater than 0 and not 7")
+		    .warnIf([](const double& val, std::string& message) {
+			    message = "Value is 7, are you sure this is a good value?";
+			    return val == 7;
+		    })
+		    .errorIf([](const double& val, std::string& message) {
+			    message = "Value must not be negative!";
+			    return val < 0;
+		    });
+
+		// "string_param", use a parameter error condition "EMPTY"
+		declare(string_param)
+		    .describe("meh")
+		    .warnIf([](const std::string& val, std::string& message) {
+			    message = "string is much too long";
+			    return val.size() > 100;
+		    })
+		    .errorIf(EMPTY);
+
+		// "double_map_param", validate keys and values, reject if not valid
+		declare(double_map_param).describe("6-dof joint state").errorIf([](const auto& values, std::string& message) {
+			message = "Invalid joint state map values, expected 6-dof";
+			return values.size() != 6 || checkJointNamesInRobotModel(values);  // from somewhere
```

Comment:
depends... Of course this example is not meant to make any actual sense, but technically parameters are already set to their overrides when you call `declare()`. That means, you could first create the RobotModel from your srdf parameter and then declare the map with proper validity checks.

Reply:
Though usually you would want to declare all parameters in the same code block (as in the `NodeConfig` class you want to provide), so entwined parsing and parameter interpretation will yield clumsy code again.
I don't have a proposal on how to improve that though.

Reply:
True! But would there always be only one config per node? Plugins might have their own configs with defaults already initialized. Even if this example (joint names) is really not perfect, I definitely see the use case of declaring parameters (plugin configs) based on other parameters (plugin class name).

Reply:
Makes sense. I guess plugins should always be forced to have their parameters in dedicated namespace then, to avoid potential clashes because of redeclared parameters?